### DISCUSSION TITLE
Issues with streaming data in textcat pipeline

### DISCUSSION THREAD

**Discussion by benbatorsky-ciox** [explosion/spaCy#7148] [Help: Coding & Implementations]
Hi,
  I've been struggling a bit trying to write a process for streaming data into the training pipeline for a text categorization component.  I'm adapting the [example in the docs for creating a generator](https://spacy.io/usage/training#custom-code-readers-batchers).  As far as I can tell, the initialization process runs through ALL examples as part of the validation step, which itself takes a while, but the major issue is that the [training loop](https://github.com/explosion/spaCy/blob/4188beda871b1e40eb8d02b8a787b6878c89717e/spacy/training/loop.py#L294) seems to be loading ALL data into memory before batching.

  I tried a workaround where I just pass the generator to the batcher, but still running into issues (it seems to stop during the first epoch, even with patience turned off).  Are there examples for how to use a generator with one of these pipelines? Are there any suggestions for ways I might be able to run this without loading everything in memory?

  Appreciate all ya'll do!

- Ben

**Comment by adrianeboyd** [comment]
Yes, the current training loop with `spacy train` doesn't work with an infinite generator. We are looking into how we can make this work again.

You'll still need to handle the label initialization separately (similar to [`spacy init labels`](https://spacy.io/api/cli#init-labels) with a finite dataset) so that the `initialize` step doesn't try to go through all your training examples to figure out which labels to add for each component.

**Reply by benbatorsky-ciox** [reply]
Thanks for the response! A couple follow-up questions and I'm happy to try and help out with either of these tweaks if that'd be useful:

1) Re: the init labels, do you have an example on hand how that works? It seems like it outputs a json, but also seems to expect a json formatted set of labels? Is there a reason why retrieving just one Example wouldn't be sufficient, since it seems the Example contains all classes (at least in my implementation).

2) Re: Infinite generators, I tried a work-around that is super simplistic, just batching the iterator (rather than unpacking it as a list.  But it doesn't seem to get through all my data and ends (without error) abruptly.  Any thoughts on why that might be?:

```
examples = iterator
#examples = list(iterator) # removed the unpacking
if not examples:
    # Raise error if no data
    raise ValueError(Errors.E986)
while max_epochs < 1 or epoch != max_epochs:
    #random.shuffle(examples) # can't shuffle everything in this case
    for batch in batcher(examples):
        yield epoch, batch
    epoch += 1
```

**Reply by adrianeboyd** [reply]
1. Run `init labels` on a finite `.spacy` dataset to see the output, which should be one `.json` file per component. One `Example` should be sufficient if it contains all the labels. Your config `[initialize]` block would then have an entry like this (adjusted for the right component):
   ```ini
   [initialize.components.textcat]
   
   [initialize.components.textcat.labels]
   @readers = "spacy.read_labels.v1"
   path = "corpus/labels/textcat.json"
   require = false
   ```
2. Hmm, if there's not an error I'm not sure, but `if not examples` going to try to check `if len(examples) > 0` internally, which isn't going to be supported for a generator. If you remove that, I think this should work. My current sketch of how to support this more generally is here, but it's not a formal PR yet (needs more tests / testing): https://github.com/adrianeboyd/spaCy/tree/feature/infinite-train-corpus

**Reply by benbatorsky-ciox** [reply]
Ahh, I see, so you need to put it in the `.spacy` format first.  Is there a way to skip that step? My generator right now just jumps from raw data to `Examples`.

Also, thanks for the link! This is similar (but more elegant) to what I've landed on.  Realized my issue was the `create_train_batches` function takes an already initialized generator, so it would just empty out in the first epoch.  Do you run into that issue with yours? It may just be the way my generator is designed.

Interestingly, `create_evaluation_callback` doesn't initialize the generator, and so each time the evaluation is done on the full dev corpus.  Seems like the same approach might solve the above issue.  Let me know your thoughts:

https://github.com/explosion/spaCy/blob/a59f3fcf5dab3acf5570483cc314b47cc5833f39/spacy/training/loop.py#L81

**Reply by adrianeboyd** [reply]
Well, for `init labels` the main thing is that you want to pass in a **finite** training corpus.. If you're using our corpus reader, you'd want to have `.spacy` files, but you can replace this with your own reader/format. I think it's likely that with an infinite training corpus, you'd need to generate the labels with a different config than you use for the final training, which is maybe a bit suboptimal, but you can also save/generate the labels files separately and reuse them as long as the labels stay the same.

The branch linked above above accumulates the examples in the first epoch so that it has a saved copy for future epochs after the generator is exhausted. An infinite generator would always stay in the first epoch and the examples should only be saved for use in future epochs if `max_epoch != 1`. This does allow bad memory leaks if your config has the wrong settings, but other alternatives complicate the config more. (Suggestions are welcome!)

The `Example` objects may potentially save some state from previous runs (although now we're copying them internally in `nlp.evaluate` to try to avoid this), so the evaluation always reloads the corpus so you're sure it's evaluating from a clean slate. It's slightly slower but the risk of unintentional side effects is much lower.

**Reply by benbatorsky-ciox** [reply]
Good point! Maybe I'm mixing things up: My generator is finite, but my corpus is too large to fit in memory.  So any operation that requires that blows up my machine.  Seems like you're suggesting a separate "label corpus"? That sounds like it would work, I'll try that out.

Yeah, and the issue you raise makes sense with an infinite generator.  My goal was to get the training to run through the whole training corpus each epoch, but avoiding loading everything in memory.  Seems like your approach would still (eventually) hold everything in that examples object? But then the shuffling would make more sense, the way I have it it'll only shuffle the batch.

**Reply by adrianeboyd** [reply]
A separate label corpus would be fine.

The case with a finite corpus that doesn't fit in memory is slightly different that what I was thinking of, but I think you'd have to handle it similarly to an infinite corpus because `spacy train` can't handle the shuffling.

I think to use `spacy train` with this you'd have to write a custom reader that handles the shuffling and repeating however it wants and presents itself to `spacy train` as if it's one huge epoch with `max_epochs = 1`. You'd control how long it trains with `max_steps` + `patience`, or by having the reader terminate at some point.

**Reply by benbatorsky-ciox** [reply]
This is super useful! Thanks for this, I'm implementing along those lines.
